[cli] Bootstrap authentication from `create-toolpad-app`

bharatkashyap opened the pull request:
- Closes #3861 

### Adds
- Authentication prompt questions to `create-toolpad-app`
<img width="1000" alt="Screenshot 2024-07-28 at 7 32 15 PM" src="https://github.com/user-attachments/assets/c3c78e33-5e97-44f7-9f02-1f1445cf6fa0">

### To Do
- [x] Bootstrap `.env.local` with empty secrets but valid keys
- [x] Handle`nextjs-pages` router branch 

### Questions
- [x] Instead of using strings to bootstrap files, create template folders for different scenarios and copy them based on user selection, instead of writing string content into files? This is [what `create-next-app` does](https://github.com/vercel/next.js/tree/canary/packages/create-next-app/templates), and it seems to be more readable and maintanable.
   - Keep using strings for now; it might scale better versus creating separate folder for each permutation of initial selection, and plays better with the webcontainers demo

Janpot commented on packages/create-toolpad-app/src/templates/dashboardLayout.ts:
```diff
@@ -0,0 +1,12 @@
+export default `import * as React from 'react';
```
I could see us make each of those into a common shape like:

```tsx
interface File {
  content: string
}

type Template = (option: Options) => File
```

That way a template has input, and can also return metadata beyond content

bharatkashyap commented:
### Update

- Added information about `.env.local` being created when auth options are selected, linking to Auth.js docs
![Screenshot 2024-08-12 at 6 07 48 PM](https://github.com/user-attachments/assets/38356b09-62eb-4281-8ee5-235a8157c078)


- Added tests to check for auth options

Janpot commented on packages/create-toolpad-app/src/templates/auth/nextjs-app/signInPage.ts:
```diff
@@ -0,0 +1,51 @@
+          if (error instanceof AuthError) {
+            return {
+              error:
+                error.type === 'CredentialsSignin'
```
If you don't have the credentials provider, this code path will never be taken

Janpot commented on packages/create-toolpad-app/src/generateProject.ts:
```diff
@@ -1,363 +1,171 @@
+import oAuthProvider from './templates/auth/oAuthProvider';
+import providerSetup from './templates/auth/providerSetup';
+import providerMap from './templates/auth/providerMap';
+import authImport from './templates/auth/import';
```
IMO this goes way too far. Totally incomprehensible to figure out what's going on in an output file without jumping to three different source files.

Janpot commented on packages/create-toolpad-app/src/templates/auth/envLocal.ts:
```diff
@@ -0,0 +1,22 @@
+${oAuthProviders
+  .map(
+    (provider) => `
+${provider?.toUpperCase()}_CLIENT_ID=
```
Only null-coalesce things when it makes sense. This one can't be null according to the types. And, if it could be `null` for some reason, that would result in `"null_CLIENT_ID"` which would be a bug anyway.


```suggestion
${provider.toUpperCase()}_CLIENT_ID=
```

Besides, you'd probably want to convert to constant case, to avoid `"azure-devops"` from turning into `"AZURE-DEVOPS"`

```tsx
// This function can go in @toolpad/utils/strings
function kebabToConstantCase(input: string): string {
  return input.split('-').map(part => part.toUpperCase()).join('_')
}
```

Janpot commented on packages/create-toolpad-app/src/types.ts:
```diff
@@ -0,0 +1,13 @@
+import { PackageJson } from './templates/packageType';
+
+export type SupportedAuthProvider = 'Credentials' | 'Google' | 'GitHub' | 'Facebook';
```
Would it make sense to just use the same ids that auth.js uses?

Janpot commented on packages/create-toolpad-app/src/templates/auth/auth.ts:
```diff
@@ -0,0 +1,74 @@
+  const providers: Provider[] = [
+    ${oAuthProviders
+      .map(
+        (provider) => `${provider}({
```
You could branch here on whether it's the credentials provider to avoid having to traverse the provider array twice in
```tsx
const hasCredentialsProvider = providers.includes('Credentials');
const oAuthProviders = providers.filter((provider) => provider !== 'Credentials');
```

Janpot commented on packages/create-toolpad-app/src/templates/nextjs-pages/indexPage.ts:
```diff
@@ -0,0 +1,25 @@
+export default function HomePage() {
+  return (
+    <Box
+      sx={{
```
TBH, if we feel the need to add any styles here, it feels like we're doing something wrong with `PageContainer`. How does it look when this is just a `<div>`?

bharatkashyap replied:
This is to be able to center the message on the screen; we can remove the styles from the `div` and instead move it to the `<Typography>`

Janpot replied:
I would just not center the message

Janpot commented on packages/create-toolpad-app/src/templates/rootPage.ts:
```diff
@@ -0,0 +1,25 @@
+const rootPage = `
+  "use client";
+  import Link from 'next/link';
+  import { Container, Typography, Box } from '@mui/material';
```
```suggestion
  import Container from '@mui/material/Container';
  import Typography from '@mui/material/Typography';
  import Box from '@mui/material/Box';
```

Janpot commented on packages/create-toolpad-app/src/templates/auth/nextjs-pages/signIn.ts:
'packages/create-toolpad-app/src/templates/auth/nextjs-pages/signIn.ts'

Just a note, but this folder structure makes it a bit harder to share things between pages router and app router. e.g. parts of the auth stuff will be identical.

bharatkashyap replied:
I've placed the parts that are identical in 'packages/create-toolpad-app/src/templates/auth/...' and then moved the router specific files to the respective folders

Janpot commented on examples/core-auth-nextjs-pages/src/pages/index.tsx:
```diff
@@ -1,21 +1,12 @@
-    >
-      <Typography variant="h4" component="h1" sx={{ mb: 2 }}>
+    <div>
+      <Typography variant="h4" component="h1" sx={{ m: 2 }}>
```
just text will be fine

```suggestion
```

Janpot commented on examples/core-auth-nextjs-pages/src/pages/orders/index.tsx:
```diff
@@ -1,21 +1,12 @@
-    >
-      <Typography variant="h4" component="h1" sx={{ mb: 2 }}>
+    <div>
+      <Typography variant="h4" component="h1" sx={{ m: 2 }}>
```
```suggestion
```
